File: 0367-valid-perfect-square/0367-valid-perfect-square.kt
class Solution {
    fun isPerfectSquare(num: Int): Boolean {
        val n = num.toLong()
        var left = 0L; var right = num.toLong()
        while(left <= right) { 
            val mid = left + (right-left) /2   
            if (mid * mid == n) return true
            else if (mid * mid < n) { 
                left = mid + 1L 
            } else {
                right = mid - 1L
            }
        } 
        return false
    }
}

File: 0367-valid-perfect-square/0367-valid-perfect-square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        lo, hi = 1, num
        while lo <= hi:
            mid = (lo + hi) // 2
            if mid * mid == num:
                return True
            elif mid * mid < num:
                lo = mid + 1
            else:
                hi = mid - 1
        return False